src/cylinder.jl

```diff
@@ -25,12 +25,21 @@ function Base.in(x::SVector{N}, s::Cylinder{N}) where {N}
     return sum(abs2,d - p*s.a) ≤ s.r^2
 end
 
-function normal(x::SVector{N}, s::Cylinder{N}) where {N}
+function surfpt_nearby(x::SVector{N}, s::Cylinder{N}) where {N}
     d = x - s.c
-    p = dot(d, s.a)
-    p > s.h2 && return s.a
-    p < -s.h2 && return -s.a
-    return normalize(d - p*s.a)
+    p = d ⋅ s.a
+    q = d - p*s.a
+    lp = abs(p)
+    lq = norm(q)
+    if abs(lp-s.h2) < abs(lq-s.r)
+        nout = sign(p)*s.a
```

Comment:
Again, avoiding the possibility of zero `nout` as you suggested is safer and a better programming practice, so I will make this change.  

However `x` with `p == 0` is very apart from the top and bottom bases, so for such `x` to be close to the surface it must be close to the side wall (in other words `abs(lp-s.h2) < abs(lq-s.r)` fails and the `if` block containing `sign(p)` is not executed).  I'm curious if there are practical cases where this is not the case.